Add support to return value in StableMIR interface and not crash due to compilation error

celinval opened the pull request:
Invoking `StableMir::run()` on a crate that has any compilation error was crashing the entire process. Instead, return a `CompilerError` so the user knows compilation did not succeed. I believe ICE will also be converted to `CompilerError`.

I'm also adding a possibility for the callback to return a value. I think it will be handy for users (at least it was for my current task of implementing a tool to validate stable-mir). However, if people disagree, I can remove that.

celinval committed:
Adjust StableMIR interface to return and not crash

Invoking StableMir::run() on a crate that has any compilation error was
crashing the entire process. Instead, return a `CompilerError` so the
user knows compilation did not succeed.

I believe ICE will also be converted to `CompilerError`.

I'm also adding a return value to the callback, because I think it will
be handy for users (at least it was for my current task of implementing
a tool to validate stable-mir). However, if people disagree,
I can remove that.

celinval commented:
r? @oli-obk @ouz-a

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6031584976/job/16365501921) [(plain)](https://github.com/rust-lang/rust/commit/9c33c5cdea46507043d31ce25d82ade7e07130d8/checks/16365501921/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=celinval
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_0a55f68d-8e41-440a-8b53-d2b0f72bae96
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c653377f3ab2ddaf11ded6f4e09556fa6197009c
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_0a55f68d-8e41-440a-8b53-d2b0f72bae96
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_0a55f68d-8e41-440a-8b53-d2b0f72bae96
GITHUB_TRIGGERING_ACTOR=celinval
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115397/merge
GITHUB_WORKFLOW_SHA=c653377f3ab2ddaf11ded6f4e09556fa6197009c
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
normalized stderr:
warning: unused `Result` that must be used
##[warning]  --> $DIR/crate-info.rs:139:5
   |
LL |     rustc_internal::StableMir::new(args, test_stable_mir).run();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this `Result` may be an `Err` variant, which should be handled
   = note: `#[warn(unused_must_use)]` on by default
   = note: `#[warn(unused_must_use)]` on by default
help: use `let _ = ...` to ignore the resulting value
   |
LL |     let _ = rustc_internal::StableMir::new(args, test_stable_mir).run();

warning: 1 warning emitted


---
--- stderr -------------------------------
warning: unused `Result` that must be used
##[warning]  --> /checkout/tests/ui-fulldeps/stable-mir/crate-info.rs:139:5
   |
LL |     rustc_internal::StableMir::new(args, test_stable_mir).run();
   |
   = note: this `Result` may be an `Err` variant, which should be handled
   = note: `#[warn(unused_must_use)]` on by default
help: use `let _ = ...` to ignore the resulting value
help: use `let _ = ...` to ignore the resulting value
   |
LL |     let _ = rustc_internal::StableMir::new(args, test_stable_mir).run();

warning: 1 warning emitted
------------------------------------------

```

</details>

ouz-a commented:
Everything looks good, but maybe `CompilerError` + some error message like "Crate you are trying to compile can't be compiled due to ICE" or something would provide better user experience(It would for me).

celinval commented:
> Everything looks good, but maybe returning `CompilerError` + some error message like "Crate you are trying to compile can't be compiled due to ICE" or something would provide better user experience(It would for me).

The compiler will still print its regular messages, so the ICE message should still print. I believe the panic hook will also run before we catch the error.

If we were to differentiate failure reasons in our APIs, we could use an enum instead of a struct, and have an ICE variant. I can do that now or we can do it later.  What do you think?

ouz-a commented:
> The compiler will still print its regular messages, so the ICE message should still print. I believe the panic hook will also run before we catch the error.
> 
> If we were to differentiate failure reasons in our APIs, we could use an enum instead of a struct, and have an ICE variant. I can do that now or we can do it later. What do you think?

Yeah compiler will definitely print out the regular ICE message but as you suggested expecting these with enum variants feels more right.

celinval committed:
Diferentiate between ICE and compilation error

celinval committed:
SMIR: Allow users to pick if compilation continues

Currently we stop compilation, but some users might want to keep going.
This is needed for us to test against rustc tests. Other tools, such as
Kani, also implements parts of their logic as a backend so it is
important for compilation to continue.

oli-obk commented:
Please add a test for these code paths, they seem easy to regress

celinval commented on compiler/rustc_smir/src/stable_mir/mod.rs:
```diff
@@ -40,6 +40,15 @@ pub type TraitDecls = Vec<TraitDef>;
 
+/// An error type used to represent an error that has already been reported by the compiler.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum CompilerError {
+    /// Internal compiler error (I.e.: Compiler crashed).
+    ICE,
+    /// Compilation failed.
+    CompilationFailed,
+}
```
In that case, what would the callback return?

oli-obk replied:
ControlFlow<T, E> seems fitting. Tho I do realize that we should even be using it inside the rustc APIs instead of rolling our own

celinval replied:
Totally. That is perfect! Thank you

celinval committed:
Add tests and use ControlFlow

bors commented:
:umbrella: The latest upstream changes (presumably #115469) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

celinval commented:
I just realized I forgot to add the new test file to git. ☺️

celinval commented:
@bors r=@oli-obk

bors commented:
:pushpin: Commit d10d8290ac50b5ea8c0ca72bf216cf414b2ada24 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk d10d8290ac50b5ea8c0ca72bf216cf414b2ada24 -->
<!-- homu: {"type":"Approved","sha":"d10d8290ac50b5ea8c0ca72bf216cf414b2ada24","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
